sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
+        self.decoder: Optional[codecs.IncrementalDecoder] = None
+
+        # Buffer data from frames belonging to the same message.
+        self.chunks: List[Data] = []
+
+        # When switching from "buffering" to "streaming", we use a thread-safe
+        # queue for transferring frames from the writing thread (library code)
+        # to the reading thread (user code). We're buffering when chunks_queue
+        # is None and streaming when it's a Queue. None is a sentinel
+        # value marking the end of the stream, superseding message_complete.
+
+        # Stream data from frames belonging to the same message.
+        self.chunks_queue: Optional[asyncio.Queue[Optional[Data]]] = None
+
+        # Flag to indicate we've paused the protocol
+        self.paused = False
+
+
+    async def get(self, timeout: Optional[float] = None) -> Optional[Data]:
+        """
+        Read the next message.
+        :meth:`get` returns a single :class:`str` or :class:`bytes`.
+        If the :message was fragmented, :meth:`get` waits until the last frame
+        is received, then it reassembles the message.
+        If ``timeout`` is set and elapses before a complete message is
+        received, :meth:`get` returns ``None``.
+        """
+        async with self.read_mutex:
+            if timeout is not None and timeout <= 0:
+                if not self.message_complete.is_set():
+                    return None
+            assert not self.get_in_progress
+            self.get_in_progress = True
+
+            # If the message_complete event isn't set yet, release the lock to
+            # allow put() to run and eventually set it.
+            # Locking with get_in_progress ensures only one thread can get here.
+            if timeout is None:
+                completed = await self.message_complete.wait()
+            elif timeout <= 0:
+                completed = self.message_complete.is_set()
+            else:
+                completed = await asyncio.wait_for(self.message_complete.wait(), timeout=timeout)
+
+            # Unpause the transport, if its paused
+            if self.paused:
+                self.protocol.resume_frames()
+                self.paused = False
+            assert self.get_in_progress
+            self.get_in_progress = False
+
+            # Waiting for a complete message timed out.
+            if not completed:
+                return None
+
+            assert self.message_complete.is_set()
+            self.message_complete.clear()
+
+            joiner: Data = b"" if self.decoder is None else ""
+            # mypy cannot figure out that chunks have the proper type.
+            message: Data = joiner.join(self.chunks)  # type: ignore
+
+            assert not self.message_fetched.is_set()
+            self.message_fetched.set()
+            self.chunks = []
+            assert self.chunks_queue is None
+
+            return message
+
+    async def get_iter(self) -> AsyncIterator[Data]:
```

Comment:
Pretty cool — I never implemented this [feature request](https://github.com/aaugustin/websockets/issues/479) in websockets itself!